Pyspark/tests/test_regular_patients.py

```diff
@@ -0,0 +1,191 @@
+    StructField('count_consec', LongType(), True),
+    StructField('max_count_consec', LongType(), True)
+    ])
+
+    df = spark.createDataFrame(data, schema)
+    df = df.withColumn('effective_from_date_formatted',
+                       f.to_date(f.col('effective_from_date_formatted')))
+    return df
+
+@pytest.fixture
+def final_result_data(spark):
+    """Test dataframe data"""
+    data = [
+        (1, True, False, False),
+        (2, False, False, False),
+        (3, False, False, False),
+        (4, False, False, False)
+        ]
+    schema = StructType([
+        StructField('patient_id', StringType()),
+        StructField('5months', BooleanType()),
+        StructField('9months', BooleanType()),
+        StructField('11months', BooleanType())
+        ])
+    df = spark.createDataFrame(data, schema)
+    return df
+
+
+@pytest.fixture
+def patients_data_instance(mocker, sample_data):
+    """Fixture to create a mock instance of PatientsData.
+    Returns: PatientsData- An instance of PatientsData with the mock loaded sample data"""
+    #mocking PatientData.load_data
+    mock_load_data = mocker.patch('regular_patients.regular_patients.PatientsData.load_data')
+    mock_load_data.return_value = sample_data
+    patients_data = PatientsData('sample-path')
+    return patients_data
+
+def test_format_date(patients_data_instance, formatted_date_data):
+    """Testing PatientsData.format_date()"""
+    result_df = (patients_data_instance.format_date()
+        .select('patient_id', 'effective_from_date_formatted'))
+    assertDataFrameEqual(result_df, formatted_date_data)
+
+def test_filter_by_date_range(patients_data_instance, filtered_data):
+    """Testing PatientsData.filter_by_date_range()"""
+    result_df = patients_data_instance.format_date()
+    result_df = patients_data_instance.filter_by_date_range(START_DATE, END_DATE)
+    result_df = result_df.select('patient_id', 'effective_from_date_formatted')
+    assertDataFrameEqual(result_df, filtered_data)
+    
+def test_checking_consecutives_records(patients_data_instance, consecutive_data):
+    """Testing PatientsData.checking_consecutives_records()"""
+    result_df = patients_data_instance.format_date()
+    result_df = patients_data_instance.filter_by_date_range(START_DATE, END_DATE)
+    result_df = patients_data_instance.checking_consecutives_records()
+    result_df = result_df.select('patient_id', 'effective_from_date_formatted',
+                                 'count_consec', 'max_count_consec')
+    assertDataFrameEqual(result_df, consecutive_data)
+
+def test_get_result(patients_data_instance, final_result_data):
+    """Testing PatientsData.get_results()"""
+    result_df = patients_data_instance.format_date()
+    result_df = patients_data_instance.filter_by_date_range(START_DATE, END_DATE)
+    result_df = patients_data_instance.checking_consecutives_records()
+    result_df = patients_data_instance.get_result()
```

Comment:
All your transformations applied to internal attribute `self.df_patients`, in this test you are reassigning value of `result_df` on each line (therefore eventually `result_df` will have only the last value from line 190)
B/c you have a _chain_ of transformations and each step is a _downstream_ for the next step, and all transformations are done inside function (b/c you do not pass `df` into function params)
Like here 
![image](https://github.com/user-attachments/assets/73554016-984f-497c-9487-4af1eecc5c3a)
It makes sense to mock self.df_patients for each step, for ex - when testing `checking_consecutives_records()` you just need to set value of  `self.df_patients = filtered_data` (`after filter_by_date_range()`)